File: lc_soln/0025_reverse_nodes_in_k_group/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = head
        count = 0
        while count < k and curr is not None:
            curr = curr.next
            count += 1
        if count < k:
            return head
        curr = self.reverseKGroup(curr, k)
        while count > 0:
            tmp = head.next
            head.next = curr
            curr = head
            head = tmp
            count -= 1
        return curr


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0025_reverse_nodes_in_k_group/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution
{
    public ListNode ReverseKGroup(ListNode head, int k)
    {
        ListNode dummy = new ListNode(next: head);
        ListNode groupPrev = dummy;
        while (true)
        {
            ListNode kth = GetKth(groupPrev, k);
            if (kth is null)
            {
                break;
            }
            ListNode groupNext = kth.next;

            ListNode prev = kth.next;
            ListNode curr = groupPrev.next;
            while (curr != groupNext)
            {
                ListNode tmp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = tmp;
            }

            ListNode tmp2 = groupPrev.next;
            groupPrev.next = kth;
            groupPrev = tmp2;
        }
        return dummy.next;
    }

    ListNode GetKth(ListNode curr, int k)
    {
        while (curr is not null && k > 0)
        {
            curr = curr.next;
            k--;
        }
        return curr;
    }
}
